### PR TITLE
Feature/bai 246 add name of approving user to UI

### PR DESCRIPTION
**Pull request by jo77419**
* Unsure whether I should have added the Menu section to the ApprovalsChip or to the [uuid].tsx
* Added in functionality to remove undefined if a managerResponse isn't provided - does this ever actually happen? I added it because there was a test for it but unsure if it's actually needed
* Generally does it look OK and should I add this info anywhere else?

### CODE REVIEW

File: src/common/ApprovalsChip.tsx
Comment (ARADDCC012) [inline src/common/ApprovalsChip.tsx:20]: Having props typed as `any` makes things harder than they need to be. Does it make sense to swap out `approvals` for `managerResponse` and `reviewerResponse` as two strongly typed objects? As it stands there's a lot of properties blindly being accessed.

File: src/common/ApprovalsChip.tsx
Comment (ARADDCC012) [inline src/common/ApprovalsChip.tsx:23]: I feel like this shouldn't be necessary. Surely we know how many approvers we have, so why are we providing and then having to filter out falsy values?

File: src/common/ApprovalsChip.tsx
Comment (ARADDCC012) [inline src/common/ApprovalsChip.tsx:24]: All of these in-line calculations (`numApprovals`, `totalApprovals`, `open`) will be calculated on every single render. I tend to prefer to memoize them so that they're only recalculated when they need to be (when one of the values in the dependency array changes). E.g:
```
const numApprovals = useMemo(() => {
  approvalOutcomes.filter((e: string) => e === 'Accepted').length
}, [approvalOutcomes])
```

There is a slight overhead to using `useMemo`, but I find it's usually worth playing it safe and using it for anything mildly complex.

File: src/common/ApprovalsChip.tsx
Comment (ARADDCC012) [inline src/common/ApprovalsChip.tsx:28]: Instead of using the `Boolean` global you can just do 
```
const open = !!anchorEl
```

File: src/common/ApprovalsChip.tsx
Comment (ARADDCC012) [inline src/common/ApprovalsChip.tsx:69]: Convention for handling events is to name the handler as `handle<event description>` so in this case `handleApprovalsClicked`

File: src/common/ApprovalsChip.tsx
Comment (ARADDCC012) [inline src/common/ApprovalsChip.tsx:78]: Stack's default spacing is 0 so you can omit that prop

File: src/common/ApprovalsChip.tsx
Comment (ARADDCC012) [inline src/common/ApprovalsChip.tsx:95]: For boolean props you don't have to pass `true` or `false`. Instead you should do `<List dense>` or just `<List>` respectively.

File: src/common/ApprovalsChip.tsx
Comment (ARADDCC012) [inline src/common/ApprovalsChip.tsx:104]: The `!== undefined` isn't needed as `approvals?.managerResponse` will either be truthy or falsy which is enough in this case. Also `'' !== undefined` returns `true` and so do a lot of other things in JS, so you might get some unintended content being rendered with this check.

File: src/common/ApprovalsChip.tsx
Comment (ARADDCC012) [inline src/common/ApprovalsChip.tsx:49]: Instead of using "magic" strings we should use an enum so that we only have to manage the strings in one place. Something like:
```
enum ApprovalStates {
  ACCEPTED = 'Accepted',
  DECLINED = 'Declined',
  'NO_RESPONSE' = 'No Response',
}

if (response === ApprovalStates.ACCEPTED) {
  // ...
}
```

File: src/common/ApprovalsChip.tsx
Comment (ARADDCC012) [inline src/common/ApprovalsChip.tsx:20]:
```
import React, { useState, useMemo, useCallback, MouseEvent, ReactElement } from 'react'

import Close from '@mui/icons-material/CloseTwoTone'
import Done from '@mui/icons-material/DoneTwoTone'
import Chip from '@mui/material/Chip'
import Menu from '@mui/material/Menu'
import List from '@mui/material/List'
import ListItem from '@mui/material/ListItem'
import ListItemIcon from '@mui/material/ListItemIcon'
import ListItemText from '@mui/material/ListItemText'
import AccessTime from '@mui/icons-material/AccessTime'
import DownArrow from '@mui/icons-material/KeyboardArrowDown'
import UpArrow from '@mui/icons-material/KeyboardArrowUp'
import Stack from '@mui/material/Stack'
import useTheme from '@mui/styles/useTheme'

import { ApprovalStates } from '../../types/interfaces'
import { Theme } from '../theme'

type Approval = {
  reviewer: string
  status: ApprovalStates
}

type ApprovalsChipProps = {
  approvals: Approval[]
}

export default function ApprovalsChip({ approvals }: ApprovalsChipProps): ReactElement {
  const theme = useTheme<Theme>()
  const [anchorEl, setAnchorEl] = useState<HTMLDivElement | null>(null)

  const open = useMemo(() => !!anchorEl, [anchorEl])
  const numApprovals = useMemo(() => approvals.filter((approval) => approval.status === 'Accepted').length, [approvals])
  const totalApprovals = useMemo(() => approvals.length, [approvals])

  const backgroundColor = useMemo(() => {
    if (numApprovals === 0) {
      return theme.palette.error.main
    }
    if (numApprovals < totalApprovals) {
      return '#dc851b'
    }
    return '#4c8a4c'
  }, [numApprovals, totalApprovals, theme])

  const getRequestResponses = useCallback((approval: Approval, index: number) => {
    let Icon
    let secondaryText = ''
    const primaryText = `${approval.reviewer}`

    if (approval.status === ApprovalStates.Accepted) {
      Icon = Done
      secondaryText = 'Approved'
    } else if (approval.status === ApprovalStates.Declined) {
      Icon = Close
      secondaryText = 'Declined'
    } else if (approval.status === ApprovalStates.NoResponse) {
      Icon = AccessTime
      secondaryText = 'Awaiting response'
    }

    return (
      <ListItem key={index}>
        <ListItemIcon>
          <Icon aria-hidden />
        </ListItemIcon>
        <ListItemText primary={primaryText} secondary={secondaryText} />
      </ListItem>
    )
  }, [])

  const requestResponseListItems = useMemo(
    () => approvals.map((approval, index) => getRequestResponses(approval, index)),
    [approvals, getRequestResponses]
  )

  const handleApprovalsClicked = (event: MouseEvent<HTMLDivElement>) => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = () => {
    setAnchorEl(null)
  }

  return (
    <Stack direction='row'>
      <Chip
        sx={{ borderRadius: 1, color: 'white', height: 'auto', backgroundColor }}
        label={`Approvals ${numApprovals}/${totalApprovals}`}
        onClick={handleApprovalsClicked}
        icon={
          open ? (
            <UpArrow sx={{ color: 'white !important', pl: 1 }} />
          ) : (
            <DownArrow sx={{ color: 'white !important', pl: 1 }} />
          )
        }
        aria-controls='model-approvals-menu'
        aria-haspopup='true'
        aria-expanded={open ? 'true' : undefined}
      />
      <Menu id='model-approvals-menu' anchorEl={anchorEl} open={open} onClose={handleClose}>
        <List dense>{requestResponseListItems}</List>
      </Menu>
    </Stack>
  )
}
```

File: server/services/request.ts
Comment (ARADDCC002) [inline server/services/request.ts:3]: Is this path correct? Other paths to the model directory look like '../models/'

File: src/common/ApprovalsChip.spec.tsx
Comment (ARADDCC002) [inline src/common/ApprovalsChip.spec.tsx:46]: Can we re-add the deleted tests with the new props? Because the component has different states it would be nice to make sure other cases are displaying correctly.

**Review by ARADDCC002 — APPROVED** [review]

**Pull request merged**
